cmd/kube-controller-manager/app/storageversionmigrator.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package app
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/metadata"
+	"k8s.io/controller-manager/controller"
+	"k8s.io/kubernetes/cmd/kube-controller-manager/names"
+	"k8s.io/kubernetes/pkg/features"
+
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	clientgofeaturegate "k8s.io/client-go/features"
+	svm "k8s.io/kubernetes/pkg/controller/storageversionmigrator"
+)
+
+const (
+	ResourceVersionControllerName = "resource-version-controller"
+)
+
+func newStorageVersionMigratorControllerDescriptor() *ControllerDescriptor {
+	return &ControllerDescriptor{
+		name:     names.StorageVersionMigratorController,
+		aliases:  []string{"svm"},
+		initFunc: startSVMController,
+	}
+}
+
+func startSVMController(
+	ctx context.Context,
+	controllerContext ControllerContext,
+	controllerName string,
+) (controller.Interface, bool, error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.StorageVersionMigrator) &&
+		!clientgofeaturegate.FeatureGates().Enabled(clientgofeaturegate.StorageVersionMigrator) {
+		return nil, false, nil
+	}
+
+	config := controllerContext.ClientBuilder.ConfigOrDie(controllerName)
+	client := controllerContext.ClientBuilder.ClientOrDie(controllerName)
+	informer := controllerContext.InformerFactory.Storagemigration().V1alpha1().StorageVersionMigrations()
+
+	dynamicClient, err := dynamic.NewForConfig(config)
+	if err != nil {
+		return nil, false, err
+	}
+
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(config)
+	if err != nil {
+		return nil, false, err
+	}
+
+	garbageCollector := controllerContext.GarbageCollector
+	if garbageCollector == nil {
```

Comment:
How would we avoid starting the garbage collector when GC is disabled and instead only run the GraphBuilder portion that this controller relies upon?

Reply:
Here is the approach I used: I extracted the Graph Builder out of GG. So now when the ControllerContext is initialized, we also initialize the Graph Builder. The controllers that need it (GC and SVM in this case) have the responsibility to start "Run" and "Sync". It's implemented in such a way that if it's started by one controller, then the other won't start it again. All the cleanup activities, like closing queues, are now moved to the Graph Builder. Since the graph builder is the producer, it also gives other controllers the ability to wait when it's resyncing the monitors.